File: app/src/main/java/com/diu/yk_games/line2box/GameProfile.java
package com.diu.yk_games.line2box;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.Random;

public class GameProfile
{
    public static SharedPreferences preferences;
    public static SharedPreferences.Editor preferencesEditor;
    public String nm=preferences.getString("nm", "Noob"+(int)Math.floor(Math.random()*(900)+100));
    public Integer matchPlayed=preferences.getInt("matchPlayed",0);
    public Integer matchWinMulti=preferences.getInt("matchWinMulti",0);
    public Integer coin=preferences.getInt("coins",100);
    public Integer lvl=preferences.getInt("lvl",getLvlByCal());

    public String playerId;
    //preferences.getBoolean("needProfile",true)
    public static void setPreferences(SharedPreferences x)
    {
        preferences=x;
        preferencesEditor = preferences.edit();
    }

    public GameProfile() {}

    public GameProfile(String nm, Integer coin, Integer matchPlayed, Integer matchWinMulti) {
        this.nm = nm;
        this.coin = coin;
        this.matchPlayed = matchPlayed;
        this.matchWinMulti = matchWinMulti;
    }
    public void apply()
    {
        preferencesEditor.putString("nm",this.nm).apply();
        preferencesEditor.putInt("coins",this.coin).apply();
        preferencesEditor.putInt("matchPlayed",this.matchPlayed).apply();
        preferencesEditor.putInt("matchWinMulti",this.matchWinMulti).apply();

    }
    public void setNm(String nm) {
        this.nm = nm;
        //preferencesEditor.putString("nm",this.nm).apply();
    }
    public void setCoin(Integer coins) {
        this.coin = coins;
        //preferencesEditor.putInt("coins",this.coin).apply();
    }

    public Integer getLvlByCal()
    {
        int mul=matchWinMulti+1;
        int pld=matchPlayed+1;
        int lv= (int) Math.sqrt(mul*(mul/7.0)+pld*2);
        return lv;
    }
    public void setMatchPlayed() {
        this.matchPlayed++;
        //preferencesEditor.putInt("matchPlayed",this.matchPlayed).apply();
    }
    public void setMatchWinMulti() {
        this.matchWinMulti++;
        //preferencesEditor.putInt("matchWinMulti",this.matchWinMulti).apply();
    }
}

File: app/src/main/java/com/diu/yk_games/line2box/model/GameProfile.kt
package com.diu.yk_games.line2box.model

import android.os.Parcelable
import com.diu.yk_games.line2box.util.asMap
import com.google.firebase.database.ServerValue
import com.google.firebase.firestore.IgnoreExtraProperties
import kotlinx.parcelize.Parcelize
import kotlin.math.sqrt

@Parcelize
@IgnoreExtraProperties
data class GameProfile(
    var nm: String = "Noob" + (100..999).random(),
    var cityNm: String = "",
    var query: String = "",  // IP Address
    var matchPlayed: Int = 0,
    var matchWinMulti: Int = 0,
    var coin: Int = 100,
    var lvl: Int = 0,
    var playerId: String = "",
    var countryEmoji: String = "",
    var countryNm: String = ""
) : Parcelable {
    fun lvlByCal(): Int {
        val mul = matchWinMulti + 1
        val pld = matchPlayed + 1
        return sqrt(mul * (mul / 7.0) + pld * 2).toInt()
    }


    fun toPlayerInfo() = PlayerInfo(
        id = playerId,
        nm = nm,
        lvl = lvlByCal(),
        coin = coin,
        seenAt = System.currentTimeMillis()
    )

    fun toPlayerInfoDB() = toPlayerInfo().asMap().plus("seenAt" to ServerValue.TIMESTAMP)

    override fun toString(): String {
        return buildString {
            appendLine("=============================")
            appendLine("Line2Box User Profile")
            appendLine("=============================")
            appendLine("Player Name  : $nm")
            appendLine("City                 : ${cityNm.ifEmpty { "Not set" }}")
            appendLine("Country          : ${countryNm.ifEmpty { "Not set" }}")
            appendLine("Player ID        : ${playerId.ifEmpty { "Not assigned" }}")
            appendLine("----------------------------------------")
            appendLine("Statistics")
            appendLine("----------------------------------------")
            appendLine("Matches Played : $matchPlayed")
            appendLine("Multiplayer Win  : $matchWinMulti")
            appendLine("Coins                   : $coin")
            appendLine("Level                    : $lvl")
            appendLine("----------------------------------------")
            appendLine("IP Address          : 118.179.0.201")
            appendLine("=============================")
        }
    }
}
